File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/ElapsedTime.java
package com.mapbox.services.android.navigation.v5.navigation;

import android.support.annotation.Nullable;

import com.mapbox.services.android.navigation.v5.exception.NavigationException;

class ElapsedTime {
  private Long start = null;
  private Long end = null;

  void start() {
    start = System.nanoTime();
  }

  @Nullable
  Long getStart() {
    return start;
  }

  void end() {
    if (start == null) {
      throw new NavigationException("Must call start() before calling end()");
    }
    end = System.nanoTime();
  }

  @Nullable
  Long getEnd() {
    return end;
  }

  long getElapsedTime() {
    if (start == null || end == null) {
      throw new NavigationException("Must call start() and end() before calling getElapsedTime()");
    }
    return end - start;
  }
}

File: libnavigation-util/src/main/java/com/mapbox/navigation/utils/time/ElapsedTime.kt
package com.mapbox.navigation.utils.time

import com.mapbox.navigation.exception.NavigationException
import com.mapbox.navigation.util.Time
import com.mapbox.navigation.utils.extensions.ifNonNull
import kotlin.math.roundToLong

class ElapsedTime(
    private val timeProvider: Time = Time.SystemImpl
) {
    var start: Long? = null
        private set
    var end: Long? = null
        private set

    companion object {
        private const val ELAPSED_TIME_DENOMINATOR = 1e+9
        private const val PRECISION = 100.0
    }

    val elapsedTime: Double
        get() {
            return ifNonNull(start, end) { startTime, endTime ->
                val elapsedTimeInNanoseconds = endTime - startTime
                val elapsedTimeInSeconds = elapsedTimeInNanoseconds / ELAPSED_TIME_DENOMINATOR
                (elapsedTimeInSeconds * PRECISION).roundToLong() / PRECISION
            } ?: throw NavigationException("Must call start() and end() before calling getElapsedTime()")
        }

    fun start() {
        start = timeProvider.nanoTime()
    }

    fun end() {
        if (start == null) {
            throw NavigationException("Must call start() before calling end()")
        }
        end = timeProvider.nanoTime()
    }
}
